[SILGen] Always serialize witness tables in non-resilient modules

jrose-apple opened the pull request:
...even if the conforming nominal type is resilient. It's the owner of the conformance whose resilience matters.

I also factored this part out into a separate check at the AST level so we can tweak it, and also so I can use it to (slightly) speed up compiling a resilient swiftinterface.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please benchmark

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
Failure seems unrelated.

@swift-ci Please test macOS

slavapestov commented on lib/AST/ProtocolConformance.cpp:
```diff
@@ -387,6 +387,23 @@ bool NormalProtocolConformance::isSynthesizedNonUnique() const {
+  // FIXME: Looking at the type is not the right long-term solution. We need an
+  // explicit mechanism for declaring conformances as 'fragile', or even
+  // individual witnesses.
+  if (!getType()->getAnyNominal()->isFormallyResilient())
```
Yeah, I think any change at the SIL level and below should use isResilient() and not isFormallyResilient().

jrose-apple replied:
This isn't at the SIL level, though. I'm about to use it for an AST thing.

jrose-apple committed:
[SILGen] Always serialize witness tables in non-resilient modules

...even if the conforming nominal type is resilient. It's the owner
of the conformance whose resilience matters.

I also factored this part out into a separate check at the AST level
so we can tweak it, and also so I can use it to (slightly) speed up
compiling a resilient swiftinterface.

jrose-apple commented:
I'd still like to run the source compat suite, if it's fixed soon, but meanwhile…

@swift-ci Please smoke test

jrose-apple commented:
@swift-ci Please test source compatibility
